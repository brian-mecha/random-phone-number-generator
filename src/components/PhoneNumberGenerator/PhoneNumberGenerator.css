@import '~antd/dist/antd.css';

.form {
  padding: 10px;
}

.download-button {
  float: right;
  padding-bottom: 20px;
}
