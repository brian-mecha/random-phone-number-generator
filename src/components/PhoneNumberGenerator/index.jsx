import React, { Component } from 'react';
import { Row, Col, Form, Input, Button, Icon, Table, Radio, Divider, Card } from 'antd';
import CsvDownloader from 'react-csv-downloader';
import './PhoneNumberGenerator.css'
import generatePhoneNumbers from '../../helpers/generateRandomNumber';

export default class PhoneNumberGenerator extends Component {
  state = {
    sortOrder: 'ASC',
    phoneNumbers: [],
    phoneNumberCount: 100,
    isDisabled: false,
    isLoading: false
  }

  radioOptions = [
    { label: 'ASC', value: 'ASC' },
    { label: 'DESC', value: 'DESC' }
  ];

  generateNewPhoneNumbers = () => {
    const { phoneNumbers, phoneNumberCount, sortOrder } = this.state;

    this.setState({ isLoading: true }, () => {
      const newNumbers = generatePhoneNumbers(phoneNumbers, phoneNumberCount, sortOrder);
      this.getNewPhoneNumbers(newNumbers);
    });
  }

  getNewPhoneNumbers = (data) => {
    this.setState({
      phoneNumbers: data,
      isLoading: false,
    });
  }

  handleRadioClick = ({ target: { value } }) => {
    this.setState({ sortOrder: value });
  }

  handleChange = ({ target: { value } }) => {
    if (value.length < 1 || value > 10000) {
      this.setState({ isDisabled: true });
    } else if (value.length >= 1) {
      this.setState({ isDisabled: false });
    }

    this.setState({ phoneNumberCount: value });
  }

  renderDownloadButton = () => {
    const { phoneNumbers } = this.state;

    const columns = [
      {
        id: 'id',
        displayName: 'No.'
      },
      {
        id: 'num',
        displayName: 'Phone Numbers'
      }
    ];

    const numbers = phoneNumbers.map((phoneNumber, index) => ({
      id: index + 1,
      num: phoneNumber.toString(),
    }));

    return (
      <CsvDownloader
        datas={numbers}
        filename="new-phone-numbers"
        columns={columns}
      />
    );
  }

  getMinMaxValue = (numbers, type) => {
    if ( numbers.length < 1) return;

    if (type === 'min') {
      return `0${Math.min(...numbers)}`;
    } else if (type === 'max') {
      return `0${Math.max(...numbers)}`;
    }
  }

  render() {
    const { isDisabled, isLoading, phoneNumberCount, phoneNumbers, sortOrder } = this.state;

    return (
      <React.Fragment>
        <h1>Phone Number Generator</h1>
        <Row>
          <Col span={24}>
            <Form className="form" layout="inline">
              <Form.Item>
                <Input placeholder="Phone number count" type="number" value ={phoneNumberCount} onChange={this.handleChange}/>
              </Form.Item>

              <Form.Item>
                <Radio.Group
                  options={this.radioOptions}
                  onChange={this.handleRadioClick}
                  value={this.state.sortOrder}
                />
              </Form.Item>

              <Form.Item>
                <Button type="primary" onClick={this.generateNewPhoneNumbers} disabled={isDisabled} loading={isLoading}>
                  <Icon type="sliders" /> Generate
                </Button>
              </Form.Item>
            </Form>

            <Divider />

            <Row>
              <Col span={24}>
              <div className="download-button">
                    {this.renderDownloadButton()}
                  </div>
              </Col>
            </Row>

            <Row type="flex" justify="space-around">
              <Col span={16}>
                <Table columns={[{title: 'Phone Numbers'}]} dataSource={phoneNumbers} bordered />
              </Col>

              <Col span={6}>
                <Card title="Statistics" bordered={false} >
                  <dl>
                    <dt><strong>Phone Numbers Count:</strong></dt>
                    <dd>{phoneNumberCount || '--'}</dd>
                    <dt><strong>Order:</strong></dt>
                    <dd>{sortOrder || '--'}</dd>
                    <dt><strong>Minimum:</strong></dt>
                    <dd>{this.getMinMaxValue(phoneNumbers, 'min') || '--'}</dd>
                    <dt><strong>Maximum:</strong></dt>
                    <dd>{this.getMinMaxValue(phoneNumbers, 'max') || '--'}</dd>
                  </dl>
                </Card>
              </Col>
            </Row>

          </Col>
        </Row>
      </React.Fragment>
    )
  }
}
