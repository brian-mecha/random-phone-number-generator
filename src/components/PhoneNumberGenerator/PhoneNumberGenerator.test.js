import React from 'react';
import { shallow, mount } from 'enzyme';

import PhoneNumberGenerator from './';

describe('Phone Number generator', () => {
  const props = {
    handleChange: jest.fn(),
    generateNewPhoneNumbers: jest.fn()
  }
  const wrapper = mount(<PhoneNumberGenerator {...props}/>)

  it('renders PhoneNumberGenerator properly', () => {
    const tree = shallow(
      <PhoneNumberGenerator {...props} />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('simulates change on the phone numbers count input box', () => {
    const input = wrapper.find('input[type="number"]');
    expect(input.instance().value).toEqual('100');

    input.simulate('change', { target: { value: 2000 }});
    expect(input.instance().value).toEqual('2000');
  });

  it('simulates change on the sort order radio buttons', () => {
    const radio = wrapper.find('input[type="radio"]');
    expect(radio.first().instance().value).toEqual('ASC');

    wrapper.find('.ant-radio-input').last().simulate('change');
    expect(wrapper.instance().state.sortOrder).toEqual('DESC');
  });

  it('simulates a click on the generate button', () => {
    const button = wrapper.find('#generate-button').last();
    const table = wrapper.find('.ant-empty');
    
    expect(button.length).toBe(1);
    button.simulate('click');
    expect(table.length).toBe(1);
  });
});
