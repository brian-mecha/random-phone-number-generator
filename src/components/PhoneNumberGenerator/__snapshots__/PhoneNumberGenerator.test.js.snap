// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone Number generator renders PhoneNumberGenerator properly 1`] = `ShallowWrapper {}`;
